# Contingent Convergence Pipeline - Scripts accompanying Whiting and Fraser, 2019 - *Contingent Convergence: The ability to detect convergent genomic evolution is dependent on population size and migration*
Full pipeline employing SLiM to investigate the effects of variable demographic parameters on genetic measures of population differentiation in adaptation and convergence

## Issues
Please report any issues to j.whiting2@exeter.ac.uk or post in the Issues tab

## SLiM 3 software
This pipeline uses the forward simulation software SLiM 3 (Haller and Messer 2019) to perform all simulations. 

SLiM can be downloaded [here](https://messerlab.org/slim/).

This pipeline employs the command-line version of SLiM 3, however a GUI is provided which aids in interpretation and visualising simulations.

## To run
**Data processing**
* Download the entire folder to your system using the `Clone` or `download` button.
* Ensure the folder structure and folder names are unchanged.
* Prior to running the pipeline, build a library of genomic windows to analyse using a GFF with (`scripts/build_a_SLIM_library.sh`)
* The pipeline is run through a single script (`scripts/slim_master_run.sh`).
* There are variants to `scripts/slim_master_run.sh`. `scripts/slim_master_run_NEUTRAL.sh` is used to simulate neutral windows under equivalent demographic treatments, and `scripts/slim_master_run_NO_EXPANSION.sh` limits the total N of individuals to 1000.
* Edit the variables at the top of the script to include the working directory, population size, mutation rate, and output directory.

**The pipeline**
* The pipeline runs by modifying the SLiM template txt files in the scripts directory to create full SLiM scripts to run.
* For each iteration of the simulation, a directory is made in which all SLiM scripts are written and temporary files are stored.
* Iternation numbers correspond to gene IDs made by `scripts/build_a_SLIM_library.sh`. These genes are read in to make scripts.
* Firstly, a burn-in script is created and run to generate an ancestral population with genetic and phenotypic variation.
* Then, the 16 demographic treatment scripts are written and run using GNU parallel for divergent phenotypes (Pheno<sub>Div</sub>).
* Then, the 16 demographic treatment scripts are written and run using GNU parallel for common phenotypes (Pheno<sub>Null</sub>).
* For each gene, for each treatment, 20 iterations are run and the results are concatenated and marked with iteration IDs.
* Outputs are combined to a single file which is exported to the user-defined output directory.
* The temporary directory for each gene is removed following final output.

* For analysis purposes, neutral data is also required. This is done in much the same way, with the script `scripts/slim_master_NEUTRAL.sh`.

**Analysis**
* All analyses, outputs and figures are generated by the `R/analyse_slim_outputs.R script`.
* This script takes all pipeline outputs from the user-defined output directory, combines them, and performs the full analysis.
* Variants of this script are used to analyse Pheno<sub>Null</sub> and Neutral simulation results separately.

## References
Haller, B. C., & Messer, P. W. (2019). SLiM 3: Forward genetic simulations beyond the Wrightâ€“Fisher model. Molecular biology and evolution, 36(3), 632-637.
